File: src/E3DCopilot.Core/Messaging/MessageContracts.cs
using System.Collections.Generic;
using Newtonsoft.Json;

namespace E3DCopilot.Core.Messaging
{
    /// <summary>
    /// 前后端消息契约定义
    /// 所有消息类型、常量、数据结构在此统一定义
    /// </summary>
    
    #region 消息类型常量
    
    public static class MessageTypes
    {
        // === 前端 → 后端 ===
        public const string UserMessage = "user:message";
        public const string UserCancel = "user:cancel";
        public const string UserNewSession = "user:new_session";
        public const string UserApprove = "user:approve";
        public const string UserAskResponse = "user:ask_response";
        public const string UserSetPlanMode = "user:set_plan_mode";
        public const string UserSetApprovalMode = "user:set_approval_mode";
        public const string UserCloseTab = "tab:close";
        public const string UserSteer = "user:steer";  // 中途干预：用户运行中注入引导消息
        public const string Ping = "ping";

        // === Provider / Model 管理（参考 Reasonix） ===
        public const string ModelsList = "models:list";
        public const string ModelSwitch = "model:switch";
        public const string ProvidersList = "providers:list";
        public const string ProviderSave = "provider:save";
        public const string ProviderDelete = "provider:delete";
        public const string ProviderFetchModels = "provider:fetch_models";
        public const string ProviderSetKey = "provider:set_key";

        // === Skills 管理 ===
        public const string SkillsList = "skills:list";
        public const string SkillsToggle = "skills:toggle";
        public const string SkillsAddSource = "skills:add_source";
        public const string SkillsRemoveSource = "skills:remove_source";
        public const string SkillsRefresh = "skills:refresh";

        // === Memory 管理 ===
        public const string MemoryList = "memory:list";
        public const string MemorySave = "memory:save";
        public const string MemoryDelete = "memory:delete";

        // === Settings 管理 ===
        public const string SettingsSave = "settings:save";

        // === Sessions 管理 ===
        public const string SessionsList = "sessions:list";
        public const string SessionsDelete = "sessions:delete";

        // === 后端 → 前端 ===
        public const string Pong = "pong";
        public const string LlmStreamDelta = "llm:stream:delta";
        public const string LlmStreamEnd = "llm:stream:end";
        public const string LlmThinking = "llm:thinking";
        public const string ToolDispatch = "tool:dispatch";
        public const string ToolResult = "tool:result";
        public const string ToolError = "tool:error";
        public const string ToolApproval = "tool:approval";
        public const string AskRequest = "ask_request";  //（对齐 Reasonix AskRequest）
        public const string Notice = "notice";
        public const string Error = "error";
        public const string HostReady = "host:ready";
        public const string ConfigSync = "config:sync";
        public const string ModelsListResult = "models:list:result";
        public const string ProvidersListResult = "providers:list:result";
        public const string ProviderFetchResult = "provider:fetch_models:result";
        public const string TurnDone = "turn:done";

        // === 新增事件类型（补全前后端事件覆盖） ===
        public const string LlmTurnStarted = "llm:turn_started";
        public const string LlmUsage = "llm:usage";
        public const string LlmRetry = "llm:retry";
        public const string ToolProgress = "tool:progress";

        // === CAD 导入相关 ===
        public const string UserCadImport = "user:cad_import";
        public const string CadProgress = "cad:progress";
        public const string CadResult = "cad:result";
        public const string CadPreview = "cad:preview";
        public const string CadConfirmBatch = "cad:confirm_batch";
        public const string CadCancel = "cad:cancel";
    }
    
    #endregion
    
    #region 前端 → 后端消息
    
    /// <summary>
    /// 用户消息（聊天输入）
    /// </summary>
    public class UserMessagePayload
    {
        [JsonProperty("text")]
        public string Text { get; set; } = "";
        
        [JsonProperty("images")]
        public string[] Images { get; set; }
        
        [JsonProperty("files")]
        public string[] Files { get; set; }

        [JsonProperty("tabId")]
        public string TabId { get; set; }
    }
    
    /// <summary>
    /// 审批响应
    /// </summary>
    public class ApprovalPayload
    {
        [JsonProperty("id")]
        public string Id { get; set; } = "";
        
        [JsonProperty("allow")]
        public bool Allow { get; set; }
    }
    
    /// <summary>
    /// 用户回答（响应 ask / ask_user，新协议对齐 Reasonix AnswerQuestion）
    /// </summary>
    public class AskResponsePayload
    {
        /// <summary>Ask 批次 ID（新协议）</summary>
        [JsonProperty("id")]
        public string Id { get; set; }

        /// <summary>问题 ID（旧协议兼容）</summary>
        [JsonProperty("questionId")]
        public string QuestionId { get; set; } = "";
        
        /// <summary>旧协议：单个回答字符串</summary>
        [JsonProperty("answer")]
        public string Answer { get; set; } = "";

        /// <summary>新协议：结构化回答列表 [{questionId, selected:[]}]</summary>
        [JsonProperty("answers")]
        public List<AskAnswerItem> Answers { get; set; }
    }

    /// <summary>
    /// 单个问题的回答（新协议，对齐 Reasonix QuestionAnswer）
    /// </summary>
    public class AskAnswerItem
    {
        [JsonProperty("questionId")]
        public string QuestionId { get; set; }
        
        [JsonProperty("selected")]
        public List<string> Selected { get; set; }
    }

    /// <summary>
    /// Plan/Act 模式切换请求
    /// </summary>
    public class SetPlanModePayload
    {
        [JsonProperty("mode")]
        public string Mode { get; set; } = "act"; // "plan" | "act"
    }

    /// <summary>
    /// Plan Mode 状态变更通知
    /// </summary>
    public class PlanModeChangedPayload
    {
        [JsonProperty("enabled")]
        public bool Enabled { get; set; }
    }
    
    #endregion
    
    #region 后端 → 前端消息
    
    /// <summary>
    /// LLM 流式输出增量
    /// </summary>
    public class LlmStreamDeltaPayload
    {
        [JsonProperty("delta")]
        public string Delta { get; set; } = "";

        [JsonProperty("tabId")]
        public string TabId { get; set; }
    }
    
    /// <summary>
    /// LLM 流式输出结束
    /// </summary>
    public class LlmStreamEndPayload
    {
        [JsonProperty("usage")]
        public object Usage { get; set; }
        
        [JsonProperty("error", NullValueHandling = NullValueHandling.Ignore)]
        public string Error { get; set; }
    }
    
    /// <summary>
    /// 工具调度
    /// </summary>
    public class ToolDispatchPayload
    {
        [JsonProperty("id")]
        public string Id { get; set; } = "";
        
        [JsonProperty("name")]
        public string Name { get; set; } = "";
        
        [JsonProperty("args")]
        public object Args { get; set; }

        [JsonProperty("tabId")]
        public string TabId { get; set; }
    }
    
    /// <summary>
    /// 工具执行结果
    /// </summary>
    public class ToolResultPayload
    {
        [JsonProperty("id")]
        public string Id { get; set; } = "";
        
        [JsonProperty("result")]
        public string Result { get; set; }
        
        [JsonProperty("error")]
        public string Error { get; set; }

        [JsonProperty("tabId")]
        public string TabId { get; set; }
    }
    
    /// <summary>
    /// 审批请求
    /// </summary>
    public class ApprovalRequestPayload
    {
        [JsonProperty("id")]
        public string Id { get; set; } = "";
        
        [JsonProperty("name")]
        public string Name { get; set; } = "";
        
        [JsonProperty("args")]
        public string Args { get; set; }
        
        [JsonProperty("description")]
        public string Description { get; set; } = "";
    }
    
    /// <summary>
    /// 通用通知
    /// </summary>
    public class NoticePayload
    {
        [JsonProperty("text")]
        public string Text { get; set; } = "";
    }
    
    /// <summary>
    /// 错误消息
    /// </summary>
    public class ErrorPayload
    {
        [JsonProperty("message")]
        public string Message { get; set; } = "";
    }
    
    /// <summary>
    /// 宿主就绪
    /// </summary>
    public class HostReadyPayload
    {
        [JsonProperty("version")]
        public string Version { get; set; } = "";
        
        [JsonProperty("platform")]
        public string Platform { get; set; } = "";
        
        [JsonProperty("timestamp")]
        public long Timestamp { get; set; }
    }
    
    /// <summary>
    /// 配置同步
    /// </summary>
    public class ConfigSyncPayload
    {
        [JsonProperty("provider")]
        public string Provider { get; set; } = "";

        [JsonProperty("model")]
        public string Model { get; set; } = "";

        [JsonProperty("baseUrl")]
        public string BaseUrl { get; set; } = "";

        [JsonProperty("apiKey")]
        public string ApiKey { get; set; } = "";

        [JsonProperty("mode")]
        public string Mode { get; set; } = "";

        // === 多 provider 模式（参考 Reasonix） ===
        [JsonProperty("currentProvider")]
        public string CurrentProvider { get; set; } = "";

        [JsonProperty("currentModel")]
        public string CurrentModel { get; set; } = "";

        [JsonProperty("providers")]
        public ProviderInfo[] Providers { get; set; }
    }

    /// <summary>
    /// 单个 provider 信息（前后端传输格式）
    /// </summary>
    public class ProviderInfo
    {
        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("kind")]
        public string Kind { get; set; } = "openai";

        [JsonProperty("baseUrl")]
        public string BaseUrl { get; set; }

        [JsonProperty("apiKey")]
        public string ApiKey { get; set; } = "";

        [JsonProperty("keySet")]
        public bool KeySet { get; set; }

        [JsonProperty("models")]
        public string[] Models { get; set; } = new string[0];

        [JsonProperty("default")]
        public string Default { get; set; } = "";

        [JsonProperty("enabled")]
        public bool Enabled { get; set; } = true;

        [JsonProperty("builtIn")]
        public bool BuiltIn { get; set; } = false;

        [JsonProperty("contextWindow")]
        public int ContextWindow { get; set; } = 0;

        [JsonProperty("visionModels")]
        public string[] VisionModels { get; set; } = new string[0];
    }

    /// <summary>
    /// 单个模型信息（前后端传输格式）
    /// </summary>
    public class ModelInfo
    {
        [JsonProperty("ref")]
        public string Ref { get; set; }      // "provider/model"

        [JsonProperty("provider")]
        public string Provider { get; set; }

        [JsonProperty("model")]
        public string Model { get; set; }

        [JsonProperty("current")]
        public bool Current { get; set; }
    }

    /// <summary>
    /// 切换模型请求（"provider/model"）
    /// </summary>
    public class ModelSwitchPayload
    {
        [JsonProperty("ref")]
        public string Ref { get; set; }
    }

    /// <summary>
    /// Provider 列表/模型列表响应
    /// </summary>
    public class ModelsListResultPayload
    {
        [JsonProperty("models")]
        public ModelInfo[] Models { get; set; } = new ModelInfo[0];

        [JsonProperty("currentProvider")]
        public string CurrentProvider { get; set; }

        [JsonProperty("currentModel")]
        public string CurrentModel { get; set; }
    }

    public class ProvidersListResultPayload
    {
        [JsonProperty("providers")]
        public ProviderInfo[] Providers { get; set; } = new ProviderInfo[0];

        [JsonProperty("currentProvider")]
        public string CurrentProvider { get; set; }

        [JsonProperty("currentModel")]
        public string CurrentModel { get; set; }
    }

    public class ProviderFetchResultPayload
    {
        [JsonProperty("providerName")]
        public string ProviderName { get; set; }

        [JsonProperty("success")]
        public bool Success { get; set; }

        [JsonProperty("models")]
        public string[] Models { get; set; } = new string[0];

        [JsonProperty("error")]
        public string Error { get; set; }
    }

    public class ProviderSavePayload
    {
        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("kind")]
        public string Kind { get; set; } = "openai";

        [JsonProperty("baseUrl")]
        public string BaseUrl { get; set; }

        [JsonProperty("apiKey")]
        public string ApiKey { get; set; }

        [JsonProperty("models")]
        public string[] Models { get; set; } = new string[0];

        [JsonProperty("default")]
        public string Default { get; set; }

        [JsonProperty("enabled")]
        public bool Enabled { get; set; } = true;

        [JsonProperty("builtIn")]
        public bool BuiltIn { get; set; } = false;

        [JsonProperty("contextWindow")]
        public int ContextWindow { get; set; } = 0;

        [JsonProperty("visionModels")]
        public string[] VisionModels { get; set; } = new string[0];
    }

    public class ProviderDeletePayload
    {
        [JsonProperty("name")]
        public string Name { get; set; }
    }

    public class ProviderSetKeyPayload
    {
        [JsonProperty("name")]
        public string Name { get; set; }

        [JsonProperty("apiKey")]
        public string ApiKey { get; set; }
    }

    #endregion

    #region CAD 导入消息

    /// <summary>
    /// CAD 导入请求
    /// </summary>
    public class CadImportPayload
    {
        [JsonProperty("action")]
        public string Action { get; set; } = "import"; // import, preview, parse_file, parse_paths

        [JsonProperty("filePath")]
        public string FilePath { get; set; }

        [JsonProperty("pathsString")]
        public string PathsString { get; set; }

        [JsonProperty("owner")]
        public string Owner { get; set; } = "/IMPORT_ZONE";

        [JsonProperty("wallHeight")]
        public double WallHeight { get; set; } = 3000;

        [JsonProperty("wallThickness")]
        public double WallThickness { get; set; } = 200;

        [JsonProperty("autoName")]
        public bool AutoName { get; set; } = true;
    }

    /// <summary>
    /// CAD 解析进度
    /// </summary>
    public class CadProgressPayload
    {
        [JsonProperty("phase")]
        public string Phase { get; set; } = ""; // parsing, classifying, creating

        [JsonProperty("current")]
        public int Current { get; set; }

        [JsonProperty("total")]
        public int Total { get; set; }

        [JsonProperty("percentage")]
        public double Percentage => Total > 0 ? (double)Current / Total * 100 : 0;

        [JsonProperty("message")]
        public string Message { get; set; } = "";
    }

    /// <summary>
    /// CAD 导入结果
    /// </summary>
    public class CadResultPayload
    {
        [JsonProperty("success")]
        public bool Success { get; set; }

        [JsonProperty("segmentCount")]
        public int SegmentCount { get; set; }

        [JsonProperty("elementCount")]
        public int ElementCount { get; set; }

        [JsonProperty("pmlScript")]
        public string PmlScript { get; set; }

        [JsonProperty("elements")]
        public object[] Elements { get; set; }

        [JsonProperty("error")]
        public string Error { get; set; }
    }

    /// <summary>
    /// CAD 预览结果
    /// </summary>
    public class CadPreviewPayload
    {
        [JsonProperty("segmentCount")]
        public int SegmentCount { get; set; }

        [JsonProperty("elementCount")]
        public int ElementCount { get; set; }

        [JsonProperty("wallHeight")]
        public double WallHeight { get; set; }

        [JsonProperty("wallThickness")]
        public double WallThickness { get; set; }

        [JsonProperty("preview")]
        public object[] Preview { get; set; }
    }

    /// <summary>
    /// 批量确认请求
    /// </summary>
    public class CadConfirmBatchPayload
    {
        [JsonProperty("elements")]
        public object[] Elements { get; set; }

        [JsonProperty("owner")]
        public string Owner { get; set; } = "/IMPORT_ZONE";

        [JsonProperty("execute")]
        public bool Execute { get; set; } = true;
    }

    #endregion
    
    #region 消息信封
    
    /// <summary>
    /// 消息信封（统一格式）
    /// </summary>
    public class CopilotMessage<T>
    {
        [JsonProperty("type")]
        public string Type { get; set; } = "";
        
        [JsonProperty("payload")]
        public T Payload { get; set; }
    }
    
    #endregion
}

File: e3d-ui/src/services/messageContracts.ts
/**
 * E小智前后端消息契约 TypeScript 类型定义
 * 与 C# E3DCopilot.Core.Messaging.MessageContracts 对应
 *
 * 使用方式：
 * import type { UserMessagePayload, MessageTypes } from './messageContracts';
 */

// ============================================
// 消息类型常量（与 C# MessageTypes 对应）
// ============================================

export const MessageTypes = {
  // === 前端 → 后端 ===
  UserMessage: 'user:message',
  UserCancel: 'user:cancel',
  UserNewSession: 'user:new_session',
  UserApprove: 'user:approve',
  UserAskResponse: 'user:ask_response',
  UserSetPlanMode: 'user:set_plan_mode',
  Ping: 'ping',
  ModelsList: 'models:list',
  ModelSwitch: 'model:switch',
  ProvidersList: 'providers:list',
  ProviderSave: 'provider:save',
  ProviderDelete: 'provider:delete',
  ProviderFetchModels: 'provider:fetch_models',
  ProviderSetKey: 'provider:set_key',

  // === Skills 管理 ===
  SkillsList: 'skills:list',
  SkillsToggle: 'skills:toggle',
  SkillsAddSource: 'skills:add_source',
  SkillsRemoveSource: 'skills:remove_source',
  SkillsRefresh: 'skills:refresh',

  // === 后端 → 前端 ===
  Pong: 'pong',
  LlmStreamDelta: 'llm:stream:delta',
  LlmStreamEnd: 'llm:stream:end',
  LlmThinking: 'llm:thinking',
  ToolDispatch: 'tool:dispatch',
  ToolResult: 'tool:result',
  ToolError: 'tool:error',
  ToolApproval: 'tool:approval',
  AskUser: 'ask_user',
  Notice: 'notice',
  Error: 'error',
  HostReady: 'host:ready',
  ConfigSync: 'config:sync',
  ModelsListResult: 'models:list:result',
  ProvidersListResult: 'providers:list:result',
  ProviderFetchResult: 'provider:fetch_models:result',
  TurnDone: 'turn:done',

  // === 新增事件类型（补全前后端事件覆盖） ===
  LlmTurnStarted: 'llm:turn_started',     // 对应前端 api_req_started
  LlmUsage: 'llm:usage',                  // Token 用量
  LlmRetry: 'llm:retry',                  // 重试事件
  ToolProgress: 'tool:progress',           // 工具执行进度
} as const;

export type MessageType = typeof MessageTypes[keyof typeof MessageTypes];

// ============================================
// 前端 → 后端消息载荷
// ============================================

export interface UserMessagePayload {
  text: string;
  images?: string[];
  files?: string[];
  tabId?: string;
}

export interface ApprovalPayload {
  id: string;
  allow: boolean;
}

export interface AskResponsePayload {
  questionId: string;
  answer: string;
}

export interface SetPlanModePayload {
  mode: 'plan' | 'act';
}

export interface PlanModeChangedPayload {
  enabled: boolean;
}

// ============================================
// 后端 → 前端消息载荷
// ============================================

export interface LlmStreamDeltaPayload {
  delta: string;
  tabId?: string;
}

export interface LlmStreamEndPayload {
  usage?: unknown;
  tabId?: string;
}

export interface ToolDispatchPayload {
  id: string;
  name: string;
  args?: unknown;
  tabId?: string;
}

export interface ToolResultPayload {
  id: string;
  result?: string;
  error?: string;
  tabId?: string;
  durationMs?: number;
}

export interface ApprovalRequestPayload {
  id: string;
  name: string;
  args?: string;
  description: string;
}

export interface AskUserPayload {
  questionId: string;
  question: string;
  data?: unknown;
}

export interface NoticePayload {
  text: string;
}

export interface ErrorPayload {
  message: string;
}

export interface HostReadyPayload {
  version: string;
  platform: string;
  timestamp: number;
}

// [Phase 2 修复] 补全 ConfigSyncPayload 遗漏字段
// Phase 0 调研发现：原 web-ui 版本缺少 currentProvider / currentModel / providers 字段，
// 导致 config:sync 事件无法完整传递后端配置。此处已补全。
export interface ConfigSyncPayload {
  provider: string;
  model: string;
  baseUrl: string;
  apiKey: string;
  mode: string;
  /** 当前选中的 Provider 名称 */
  currentProvider: string;
  /** 当前选中的 Model 名称 */
  currentModel: string;
  /** 后端同步的 Provider 完整列表 */
  providers: ProviderInfo[];
}

// ============================================
// 消息信封
// ============================================

export interface CopilotMessage<T> {
  type: string;
  payload: T;
}

// ============================================
// 类型守卫函数
// ============================================

export function isMessageType(type: string): type is MessageType {
  return Object.values(MessageTypes).includes(type as MessageType);
}

// ============================================
// 消息创建辅助函数
// ============================================

export function createMessage<T>(type: MessageType, payload: T): CopilotMessage<T> {
  return { type, payload };
}

export function createUserMessage(text: string, images?: string[], files?: string[]): CopilotMessage<UserMessagePayload> {
  return createMessage(MessageTypes.UserMessage, { text, images, files });
}

export function createApproval(id: string, allow: boolean): CopilotMessage<ApprovalPayload> {
  return createMessage(MessageTypes.UserApprove, { id, allow });
}

export function createAskResponse(questionId: string, answer: string): CopilotMessage<AskResponsePayload> {
  return createMessage(MessageTypes.UserAskResponse, { questionId, answer });
}

// ============================================
// Provider / Model 管理（参考 Reasonix）
// ============================================

export interface ModelInfo {
  ref: string;        // "provider/model"
  provider: string;
  model: string;
  current: boolean;
}

export interface ProviderInfo {
  name: string;
  kind: string;
  baseUrl: string;
  apiKey: string;
  keySet: boolean;
  models: string[];
  default: string;
  enabled: boolean;
  builtIn: boolean;
}

export interface ModelsListResultPayload {
  models: ModelInfo[];
  currentProvider: string;
  currentModel: string;
}

export interface ProvidersListResultPayload {
  providers: ProviderInfo[];
  currentProvider: string;
  currentModel: string;
}

export interface ProviderFetchResultPayload {
  providerName: string;
  success: boolean;
  models: string[];
  error?: string;
}

export interface ModelSwitchPayload {
  ref: string; // "provider/model"
}

export interface ProviderSavePayload {
  name: string;
  kind: string;
  baseUrl: string;
  apiKey?: string;
  models: string[];
  default: string;
  enabled: boolean;
  builtIn: boolean;
}

export interface ProviderDeletePayload {
  name: string;
}

export interface ProviderSetKeyPayload {
  name: string;
  apiKey: string;
}

// ============================================
// Skills 管理
// ============================================

export interface SkillInfo {
  name: string;
  description: string;
  scope: 'builtin' | 'project' | 'global' | 'custom';
  runAs: 'inline' | 'subagent';
  enabled: boolean;
  filePath?: string;
  tags?: string[];
}

export interface SkillSource {
  path: string;
  status: 'active' | 'missing' | 'error';
  skillCount: number;
  removable: boolean;
}

export interface SkillsListResultPayload {
  skills: SkillInfo[];
  sources: SkillSource[];
}

export interface SkillsTogglePayload {
  name: string;
  enabled: boolean;
}

// ============================================
// Memory 管理
// ============================================

export interface MemoryEntry {
  id: string;
  title: string;
  content: string;
  kind: string;
  tags: string[];
  score?: number;
  created_at?: string;
  updated_at?: string;
}

export interface MemoryListResultPayload {
  memories: MemoryEntry[];
}

// ============================================
// Sessions 管理
// ============================================

export interface SessionInfo {
  id: string;
  title: string;
  messageCount: number;
  createdAt: number;
  lastActivityAt: number;
}

export interface SessionsListResultPayload {
  sessions: SessionInfo[];
}
